## Final Project - SQL

#### By: MadjiteyY

#### Goal: query the database to provide answers to questions that will be used to generate a value proposition for a new product.

In [1]:
# import libraries

import pandas as pd
from sqlalchemy import create_engine


db_config = {'user': 'praktikum_student',         # user name
             'pwd': 'Sdf4$2;d-d30pp', # password
             'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
             'port': 6432,              # connection port
             'db': 'data-analyst-final-project-db'}          # the name of the data base

connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
                                                                     db_config['pwd'],
                                                                       db_config['host'],
                                                                       db_config['port'],
                                                                       db_config['db'])

engine = create_engine(connection_string, connect_args={'sslmode':'require'})

In [2]:
# a look into the database
pd.io.sql.read_sql(sql = 'SELECT * FROM information_schema.tables;', con = engine)

table_catalog        table_schema          table_name  \
0    data-analyst-final-project-db              public              orders   
1    data-analyst-final-project-db              public              visits   
2    data-analyst-final-project-db              public  advertisment_costs   
3    data-analyst-final-project-db              public                trip   
4    data-analyst-final-project-db              public              C:oks   
..                             ...                 ...                 ...   
181  data-analyst-final-project-db  information_schema      foreign_tables   
182  data-analyst-final-project-db              public             reviews   
183  data-analyst-final-project-db              public             ratings   
184  data-analyst-final-project-db              public    finished_lessons   
185  data-analyst-final-project-db              public               books   

     table_type self_referencing_column_name reference_generation  \
0    BASE TABLE                         None                 None   
1    BASE TABLE                         None                 None   
2    BASE TABLE                         None                 None   
3    BASE TABLE                         None                 None   
4    BASE TABLE                         None                 None   
..          ...                          ...                  ...   
181        VIEW                         None                 None   
182  BASE TABLE                         None                 None   
183  BASE TABLE                         None                 None   
184  BASE TABLE                         None                 None   
185  BASE TABLE                         None                 None   

    user_defined_type_catalog user_defined_type_schema user_defined_type_name  \
0                        None                     None                   None   
1                        None                     None                   None   
2                        None                     None                   None   
3                        None                     None                   None   
4                        None                     None                   None   
..                        ...                      ...                    ...   
181                      None                     None                   None   
182                      None                     None                   None   
183                      None                     None                   None   
184                      None                     None                   None   
185                      None                     None                   None   

    is_insertable_into is_typed commit_action  
0                  YES       NO          None  
1                  YES       NO          None  
2                  YES       NO          None  
3                  YES       NO          None  
4                  YES       NO          None  
..                 ...      ...           ...  
181                 NO       NO          None  
182                YES       NO          None  
183                YES       NO          None  
184                YES       NO          None  
185                YES       NO          None  

[186 rows x 12 columns]

In [3]:
# describe tables of interest
pd.io.sql.read_sql(sql = """SELECT table_name, column_name, data_type FROM information_schema.columns WHERE table_name \
                                IN ('books', 'authors', 'publishers', 'ratings', 'reviews');""", con = engine)

table_name       column_name data_type
0   publishers      publisher_id   integer
1   publishers         publisher      text
2      authors         author_id   integer
3      authors            author      text
4      reviews         review_id   integer
5      reviews           book_id   integer
6      reviews          username      text
7      reviews              text      text
8      ratings         rating_id   integer
9      ratings           book_id   integer
10     ratings          username      text
11     ratings            rating   integer
12       books           book_id   integer
13       books         author_id   integer
14       books             title      text
15       books         num_pages   integer
16       books  publication_date      date
17       books      publisher_id   integer

In [4]:
# print first rows of books table
pd.io.sql.read_sql(sql = 'SELECT * FROM books LIMIT 5;', con = engine)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

In [5]:
# print first rows of authors table
pd.io.sql.read_sql(sql = 'SELECT * FROM authors LIMIT 5;', con = engine)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

In [6]:
# print first rows of ratings table
pd.io.sql.read_sql(sql = 'SELECT * FROM ratings LIMIT 5;', con = engine)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

In [7]:
# print first rows of publishers table
pd.io.sql.read_sql(sql = 'SELECT * FROM publishers LIMIT 5;', con = engine)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

In [8]:
# print first rows of reviews table
pd.io.sql.read_sql(sql = 'SELECT * FROM reviews LIMIT 5;', con = engine)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

### Tasks

***Find the number of books released after January 1, 2000.***

In [9]:
pd.io.sql.read_sql(sql = """SELECT COUNT(*) AS "Number of Books" FROM books WHERE publication_date > '2000-01-01'""", \
                   con = engine)

Number of Books
0              819

`819 books were released after January 1, 2000.`

***Find the number of user reviews and the average rating for each book.***

In [10]:
pd.io.sql.read_sql(sql = """SELECT b.book_id, b.title, COUNT(rs.*) AS "Number of User Reviews", \
                            AVG(r.rating) AS "Average Rating" FROM books AS "b" JOIN ratings AS "r" ON b.book_id = r.book_id \
                            JOIN reviews AS "rs" ON r.book_id = rs.book_id GROUP BY b.book_id, b.title \
                            ORDER BY "Average Rating" DESC""", con = engine)

book_id                                              title  \
0         55          A Woman of Substance (Emma Harte Saga #1)   
1        518               Pop Goes the Weasel (Alex Cross  #5)   
2        347  In the Hand of the Goddess (Song of the Liones...   
3        513               Piercing the Darkness (Darkness  #2)   
4         57                    Act of Treason (Mitch Rapp  #9)   
..       ...                                                ...   
989      915  The World Is Flat: A Brief History of the Twen...   
990      371                                              Junky   
991      316                  His Excellency: George Washington   
992      202                                      Drowning Ruth   
993      303                               Harvesting the Heart   

     Number of User Reviews  Average Rating  
0                         4            5.00  
1                         4            5.00  
2                         6            5.00  
3                         4            5.00  
4                         4            5.00  
..                      ...             ...  
989                      12            2.25  
990                       4            2.00  
991                       4            2.00  
992                       9            2.00  
993                       4            1.50  

[994 rows x 4 columns]

`We see that a book can have the maximum rating and still have the lowest number of user reviews. For instance, A woman of substance has the highest average rating but a somewhat low number of reviews.`

***Identify the publisher that has released the greatest number of books with more than 50 pages (this will help you exclude brochures and similar publications from your analysis).***

In [11]:
pd.io.sql.read_sql(sql = """SELECT p.publisher_id, p.publisher, COUNT(b.*) AS "Number of books" FROM books AS "b" \
                            JOIN publishers AS "p" ON b.publisher_id = p.publisher_id WHERE b.num_pages > 50 \
                            GROUP BY p.publisher_id, p.publisher ORDER BY "Number of books" DESC LIMIT 1;""", con = engine)

publisher_id      publisher  Number of books
0           212  Penguin Books               42

`Penguin books looks like the publisher that has the greatest nummber of books released.`

***Identify the author with the highest average book rating (look only at books with at least 50 ratings).***

In [12]:
pd.io.sql.read_sql(sql = """SELECT a.author, b.book_id, AVG(r.rating) AS "Average book rating" FROM books AS "b" \
                            JOIN authors AS "a" ON b.author_id = a.author_id JOIN ratings AS "r" ON b.book_id = r.book_id \
                            GROUP BY b.book_id, a.author HAVING COUNT(r.rating) >= 50 ORDER BY "Average book rating" DESC \
                            LIMIT 1;""", con = engine)

author  book_id  Average book rating
0  J.K. Rowling/Mary GrandPré      302             4.414634

`J.K. Rowling/Mary GrandPré top in this category with a 4.4 average book rating.`

***Find the average number of text reviews among users who rated more than 50 books.***

In [13]:
pd.io.sql.read_sql(sql = """WITH reviewers AS (SELECT username, COUNT(*) AS "Number of reviews" FROM reviews \
                            WHERE username IN (SELECT username FROM ratings GROUP BY username HAVING COUNT(rating)> 50) \
                            GROUP BY username) SELECT username, AVG("Number of reviews") OVER() AS "Average number of text \
                            reviews" FROM reviewers;""", con = engine)

username  Average number of text                             reviews
0     sfitzgerald                                          24.333333         
1  jennifermiller                                          24.333333         
2          xdavis                                          24.333333         
3          paul88                                          24.333333         
4      martinadam                                          24.333333         
5       richard89                                          24.333333

`24.3 is the average number of text reviews among users who rated more than 50 books.`